Add linux-x64 hosted tpdiff to superpmi-diffs pipeline

BruceForstall opened the pull request:
With this, we get instruction count diffs based on clang compiled native JITs.

BruceForstall committed:
Add linux-x64 hosted tpdiff to superpmi-diffs pipeline

With this, we get instruction count diffs based on clang compiled
native JITs.

BruceForstall committed:
Fix some variables

Also, fix some Python 'lint' issues.

BruceForstall committed:
Fix more path issues, Windows targeting global var

BruceForstall commented:
One downside: linux-x64 tpdiff is a little slow compared to the Windows runs. It extends the long pole about 15 minutes.

Example of linux-x64 run data:

<html>
<body>
<!--StartFragment--><p style="box-sizing: border-box; color: rgb(0, 0, 0); font-family: &quot;Segoe UI VSS (Regular)&quot;, &quot;Segoe UI&quot;, -apple-system, BlinkMacSystemFont, Roboto, &quot;Helvetica Neue&quot;, Helvetica, Ubuntu, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">All contexts:</p>

Collection | Base # instructions | Diff # instructions | PDIFF
-- | -- | -- | --
benchmarks.run_pgo.linux.x64.checked.mch | 184,652,213,418 | 184,652,105,802 | -0.00%
benchmarks.run_tiered.linux.x64.checked.mch | 34,283,602,135 | 34,283,533,308 | -0.00%
libraries.pmi.linux.x64.checked.mch | 215,101,941,472 | 215,101,662,141 | -0.00%
coreclr_tests.run.linux.x64.checked.mch | 898,577,412,336 | 898,576,744,617 | -0.00%
libraries_tests.pmi.linux.x64.checked.mch | 456,669,811,876 | 456,669,237,387 | -0.00%
libraries.crossgen2.linux.x64.checked.mch | 66,079,201,489 | 66,079,137,673 | -0.00%
benchmarks.run.linux.x64.checked.mch | 52,478,608,587 | 52,478,534,505 | -0.00%

<br class="Apple-interchange-newline"><!--EndFragment-->
</body>
</html>


Compared to the same MCH files run using the cross-compiler on Windows (not in the same order for some reason):

<html>
<body>
<!--StartFragment--><p style="box-sizing: border-box; color: rgb(0, 0, 0); font-family: &quot;Segoe UI VSS (Regular)&quot;, &quot;Segoe UI&quot;, -apple-system, BlinkMacSystemFont, Roboto, &quot;Helvetica Neue&quot;, Helvetica, Ubuntu, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">All contexts:</p>

Collection | Base # instructions | Diff # instructions | PDIFF
-- | -- | -- | --
benchmarks.run.linux.x64.checked.mch | 50,226,736,669 | 50,226,646,783 | -0.00%
benchmarks.run_pgo.linux.x64.checked.mch | 176,284,877,969 | 176,285,139,098 | +0.00%
benchmarks.run_tiered.linux.x64.checked.mch | 32,691,625,146 | 32,691,595,491 | -0.00%
coreclr_tests.run.linux.x64.checked.mch | 818,926,377,299 | 818,923,593,948 | -0.00%
libraries.crossgen2.linux.x64.checked.mch | 63,305,779,675 | 63,305,732,565 | -0.00%
libraries.pmi.linux.x64.checked.mch | 205,978,054,388 | 205,977,891,786 | -0.00%
libraries_tests.pmi.linux.x64.checked.mch | 437,128,801,827 | 437,128,019,912 | -0.00%

<br class="Apple-interchange-newline"><!--EndFragment-->
</body>
</html>

Note that the cross-compiler always runs fewer instructions, sometimes significantly fewer.

BruceForstall commented:
@jakobbotsch PTAL
cc @dotnet/jit-contrib

BruceForstall commented:
Sample diffs: https://dev.azure.com/dnceng-public/public/_build/results?buildId=268530&view=ms.vss-build-web.run-extensions-tab

jakobbotsch commented on src/coreclr/scripts/superpmi_diffs_setup.py:
```diff
@@ -211,6 +234,7 @@ def main(main_args):
     Note:
     1. asmdiffs uses Checked JITs, tpdiff uses Release JITs. Only the one needed is copied to the payload directory.
     2. Only asmdiffs needs jit-analyze and git
+    3. tpdiff can run on Linux, but asmdiffs is not implemented to run on Linux
```
What is the problem with asmdiffs?

kunalspathak replied:
Is it because of the availability of cordistools?

BruceForstall replied:
Not cordistools. It's because we need jit-analyze and `git`. In particular, we need `git` on the Helix machine so it can be used for `git diff` by jit-analyze. For Windows, we download a "portable git" package to the correlation payload. Building jit-analyze for Linux wouldn't be hard, but would require a few tweaks.

However, bottom-line: we don't need to run diffs natively on Linux. The results would be (should be!) identical to running them on Windows.

jakobbotsch replied:
One (obvious) benefit would be to allow contributors who work on Linux systems to run asmdiffs locally. I think we should consider investing in it since it doesn't sound like it would be too expensive.

jakobbotsch replied:
I opened #86181 for it

kunalspathak replied:
> allow contributors who work on Linux systems to run asmdiffs locally.

can't you already do that? I think @BruceForstall is referring to the problems we encounter when doing it in CI.

jakobbotsch replied:
Not sure? The comment seemed a bit general

> asmdiffs is not implemented to run on linux

If it's just in CI then :+1:

kunalspathak replied:
Yeah, this was in the context of CI (and @BruceForstall  can correct me). But to run `asmdiff`, you need:
1. baseline jit (we already have that for native binaries like `libclrjit.so`)
2. cordistools
3. jit-analyze, hopefully this would just work cross plat.

jakobbotsch replied:
Ah, this comment is not in superpmi.py, but rather in the pipeline python script, so `asmdiffs` here probably refers to the pipeline job and not superpmi command?

kunalspathak replied:
Correct.

jakobbotsch reviewed (approved):
Very nice.

jakobbotsch commented:
> Note that the cross-compiler always runs fewer instructions, sometimes significantly fewer.

It would be interesting to compare a detailed breakdown from Windows with one from Linux to see which methods executing are most different. (Not suggesting that you do that.)

kunalspathak commented on src/coreclr/scripts/superpmi_diffs.py:
```diff
@@ -53,9 +63,10 @@ def setup_args(args):
                         "Invalid type \"{}\"".format)
 
     coreclr_args.verify(args,
```
don't we need argument parsing for the newly renamed `host_os`?

BruceForstall replied:
`host_os` is a little magic, as it's parsed by the `coreclr_arguments.py` script we use.

kunalspathak commented on src/coreclr/scripts/superpmi_diffs.py:
```diff
@@ -105,43 +117,59 @@ def __init__(self, coreclr_args):
         # List of summary MarkDown files
         self.summary_md_files = []
 
+    def determine_jit_name(self):
```
This is the 3rd copy of this method in our scripts, other 2 being in `superpmi.py` and `jitrollingbuild.py`. Can we move this to `jitutil.py`?

BruceForstall commented on src/coreclr/scripts/superpmi-diffs.proj:
```diff
@@ -24,13 +24,27 @@
+    <ProductDirectory>$HELIX_CORRELATION_PAYLOAD</ProductDirectory>
+    <SuperpmiLogsLocation>$HELIX_WORKITEM_UPLOAD_ROOT</SuperpmiLogsLocation>
+    <!-- Workaround until https://github.com/dotnet/arcade/pull/6179 is not available -->
+    <HelixResultsDestinationDir>$(BUILD_SOURCESDIRECTORY)/artifacts/helixresults</HelixResultsDestinationDir>
```
It might be possible, but things in msbuild get evaluated at different times, and I have a very weak understanding of when. So I'm mostly following existing practice (i.e., copy/paste development).

kunalspathak replied:
You are using `$(FileSeparatorChar)` few lines below to populate the path, so it should work. Just want to simplify the proj file with as little if-else as possible.

BruceForstall replied:
I'm going to leave it. This case will go away if I ever get around to fixing https://github.com/dotnet/runtime/issues/75169. I'll remove use of FileSeparatorChar in the python paths, since Python can handle `/` properly.

BruceForstall committed:
Feedback

1. Make `determine_jit_name` a utility library function.
All 3 versions were different, so required some adjustments.
jitrollingbuild.py had not previously used `jitutil` so added use
and removed additional duplicated code.
2. Renamed partitions to not use `unix` name. This is purely an internal
name, not exposed.
3. Simplified some FileSeparatorChar usage.
4. Fixed more pylint complaints

kunalspathak commented:
After your latest changes, should we run `superpmi-diff` once to make sure that things are good to merge? I don't recall why we decided to name `unix` and `win` and if there are more changes needed.

BruceForstall commented:
> After your latest changes, should we run `superpmi-diff` once to make sure that things are good to merge? I don't recall why we decided to name `unix` and `win` and if there are more changes needed.

I already kicked off that as well as superpmi-replay and superpmi-asmdiffs-checked-release. I presume we used "unix" because it matched the JIT name, but that name in SPMI_Partition is not exposed anywhere, except possibly as the name of the Helix job log directory.

BruceForstall commented:
Latest diff run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=269449&view=results

BruceForstall commented:
@kunalspathak PTAL

kunalspathak reviewed (approved):
LGTM
